platforms/allwinner-d1/boards/lichee-rv/src/spim.rs

```diff
@@ -0,0 +1,224 @@
+        w
+    });
+    spi1.spi_fcr.modify(|_r, w| {
+        w.tf_drq_en().enable();
+        w
+    });
+
+    Spim1 { _x: () }
+}
+
+pub struct SpiSender {
+    _x: (),
+}
+
+impl SpiSender {
+    pub async fn register(kernel: &'static Kernel, queued: usize) -> Result<(), ()> {
+        let (kprod, kcons) = KChannel::new_async(kernel, queued).await.split();
+
+        kernel
+            .spawn(async move {
+                let kcons = kcons;
+                let spi = unsafe { &*SPI_DBI::PTR };
+
+                let txd_ptr: *mut u32 = spi.spi_txd.as_ptr();
+                let txd_ptr: *mut u8 = txd_ptr.cast();
+                let txd_ptr: *mut () = txd_ptr.cast();
+
+                loop {
+                    let msg: Message<SpiSender> = kcons.dequeue_async().await.unwrap();
+                    // println!("DEQUEUE");
+                    let Message { msg, reply } = msg;
+                    let SpiSenderRequest::Send(ref payload) = msg.body;
+
+                    spi.spi_bcc.modify(|_r, w| {
+                        w.stc().variant(payload.len() as u32);
+                        w
+                    });
+                    spi.spi_mbc.modify(|_r, w| {
+                        w.mbc().variant(payload.len() as u32);
+                        w
+                    });
+                    spi.spi_mtc.modify(|_r, w| {
+                        w.mwtc().variant(payload.len() as u32);
+                        w
+                    });
+
+                    spi.spi_tcr.modify(|_r, w| {
+                        w.xch().initiate_exchange();
+                        w
+                    });
+
+                    let d_cfg = DescriptorConfig {
+                        source: payload.as_ptr().cast(),
+                        destination: txd_ptr,
+                        byte_counter: payload.len(),
+                        link: None,
+                        wait_clock_cycles: 0,
+                        bmode: BModeSel::Normal,
+                        dest_width: DataWidth::Bit8,
+                        dest_addr_mode: AddressMode::IoMode,
+                        dest_block_size: BlockSize::Byte1,
+                        dest_drq_type: DestDrqType::Spi1Tx,
+                        src_data_width: DataWidth::Bit8,
+                        src_addr_mode: AddressMode::LinearMode,
+                        src_block_size: BlockSize::Byte1,
+                        src_drq_type: SrcDrqType::Dram,
+                    };
+                    let descriptor = d_cfg.try_into().map_err(drop)?;
+                    unsafe {
+                        let mut chan = Channel::summon_channel(1);
+                        chan.set_channel_modes(ChannelMode::Wait, ChannelMode::Handshake);
+                        chan.start_descriptor(NonNull::from(&descriptor));
+                    }
+                    match SPI1_TX_DONE.wait().await {
+                        Ok(_) => {}
+                        Err(_) => todo!(),
+                    }
```

Comment:
nit, take it or leave it: this could probably just be
```suggestion
                    SPI1_TX_DONE
                        .wait()
                        .expect("SPI1_TX_DONE WaitCell should never be closed");
```
since waiting on a `WaitCell` only errors if we have called `WaitCell::close`, and AFAIK we never do that here.

Reply:
I think you missed an `await`, but I agree with you